packages/hoprd/src/logs.ts

```diff
@@ -1,12 +1,44 @@
 import ws from 'ws'
 import debug from 'debug'
+import RunQueue from 'run-queue'
+import { randomBytes } from 'crypto'
+import { Ed25519Provider } from 'key-did-provider-ed25519'
+import KeyResolver from 'key-did-resolver'
+import { DID } from 'dids'
+import CeramicClient from '@ceramicnetwork/http-client'
+import { TileDocument } from '@ceramicnetwork/stream-tile'
 
 export type Socket = ws
 
 let debugLog = debug('hoprd')
 
 const MAX_MESSAGES_CACHED = 100
 
+class Queue {
+  private queue: RunQueue = undefined
+  private completed: boolean = false
+  private opts: {
+    maxConcurrency: number
+  } = undefined
+
+  constructor(opts) {
+    this.opts = opts
+    this.queue = new RunQueue(opts)
+  }
+  add(priority, job) {
+    if (this.completed) {
+      this.queue = new RunQueue(this.opts)
+      this.completed = false
+    }
+    this.queue.add(priority, job)
+  }
+  async run() {
+    this.queue.run().then(() => (this.completed = true))
```

Comment:
How is back-pressure handled? Can a slow endpoint lead to memory problems for the node?

Reply:
That's the point of the queue, to handle back-pressure.

But that's a good question, will bring that to the Ceramic team cc @oed

Reply:
@stbrody thoghts on this?

Reply:
My naive assumption would be that this can turn into a large heap allocation in the node in case of problems unless the log forwarding has tight send time limits.

Reply:
> That's the point of the queue, to handle back-pressure.
> 
> But that's a good question, will bring that to the Ceramic team cc @oed

The queue itself doesn't handle back-pressure, only ordering.

Reply:
The Ceramic node has an internal queue that processes incoming updates, with a limit on how many writes can be processed concurrently (configurable via the [concurrentRequestsLimit](https://github.com/ceramicnetwork/js-ceramic/blob/9dfc167b4be82db79a99e3f34ddefeff49516721/packages/core/src/ceramic.ts#L91) config option). Once there are more write operations than the `concurrentRequestsLimit`, subsequent write requests will queue up and build up until previous writes have completed and drained the queue. There is not currently any explicit back-pressure mechanism and nothing that will cause operations to proactively fail when the node has too many pending requests.

CC @ukstv

Reply:
Thanks for jumping in @stbrody @eod! We'll be soon having our users run these queries against your node, so we'll be putting this to the test. For now, it sounds as if we have enough in place to take our current load.